File: LeetCode75_LevelTwo/Kth_Smallest_Element_in_a_BST/Kth_Smallest_Element_in_a_BST.py
# 230. Kth Smallest Element in a BST Python
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        def helper(root) :
            if not root: return []
            nums = [root.val]
            if root.left:
                nums = helper(root.left) + nums
            if len(nums) >= k:
                return nums
            if root.right:
                nums += helper(root.right)
            return nums
        
        num = helper(root)
        return num[k-1]

File: LeetCode75_LevelTwo/Kth_Smallest_Element_in_a_BST/Kth_Smallest_Element_in_a_BST.cc
// 230. Kth Smallest Element in a BST C++
#include <vector>

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    std::vector<int> v;
public:
    int kthSmallest(TreeNode* root, int k) {
        std::vector<int> v;
        helper(root, k, v);
        return v[k-1];
    }
    void helper(TreeNode* root, int k, std::vector<int> &v) {
        if (root->left) helper(root->left, k, v);
        v.emplace_back(root->val);
        if (v.size() >= k) return;
        if(root->right) helper(root->right, k, v);
    }
};
